clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -268,6 +280,13 @@ public void poll(long timeout) throws IOException {
         currentTimeNanos = endSelect;
         this.sensors.selectTime.record(endSelect - startSelect, time.milliseconds());
 
+        // Perform setup for instantly connected channels since they won't be included in selectedKeys()
+        if (connectableChannels.size() > 0)
+            for (KafkaChannel channel : connectableChannels) {
+                sensors.maybeRegisterConnectionMetrics(channel.id());
```

Comment:
My OP_WRITE solution was an attempt to avoid duplication and side effects. It would seem that combining keys and iterating them together is the only clean solution. I am offline today and tomorrow but will give that a try on Saturday.